blockchain/validate.go

```diff
@@ -2872,9 +3289,101 @@ func (b *BlockChain) consensusScriptVerifyFlags(node *blockNode) (txscript.Scrip
 		scriptFlags |= txscript.ScriptVerifyCheckSequenceVerify
 		scriptFlags |= txscript.ScriptVerifySHA256
 	}
+
+	// Enable OP_TADD/OP_TSPEND if treasury opcodes are active.
+	isTreasuryEnabled, err := b.isTreasuryAgendaActive(node.parent)
+	if err != nil {
+		return 0, err
+	}
+	if isTreasuryEnabled {
+		scriptFlags |= txscript.ScriptVerifyTreasury
+	}
+
 	return scriptFlags, err
 }
 
+// tspendChecks verifies that a TSpend is allowed to be mined in the provided
+// block. It verifies that it is on a TVI, is within the correct window, has
+// not been mined before and that it doesn't overspend the treasury. This
+// function assumes that the treasury agenda is enabled.
+func (b *BlockChain) tspendChecks(prevNode *blockNode, block *dcrutil.Block) error {
+	blockHeight := prevNode.height + 1
+
+	isTVI := standalone.IsTreasuryVoteInterval(uint64(block.Height()),
+		b.chainParams.TreasuryVoteInterval)
+
+	tspends := make([]*dcrutil.Tx, 0, len(block.STransactions()))
+	for _, stx := range block.STransactions() {
+		if !stake.IsTSpend(stx.MsgTx()) {
+			continue
+		}
+		// If block is not TVI error out since it is not
+		// allowed to contain a TSpend. This check is also
+		// performed in checkBlockPositional.
+		if !isTVI {
+			str := fmt.Sprintf("block contains TSpend "+
+				"transaction (%v) while block is not "+
+				"at a TVI height (%v)", stx.Hash(),
+				block.Height())
+			return ruleError(ErrNotTVI, str)
+		}
+
+		// Assert that the tspend is inside the correct window.
+		exp := stx.MsgTx().Expiry
+		if !standalone.InsideTSpendWindow(blockHeight,
+			exp, b.chainParams.TreasuryVoteInterval,
+			b.chainParams.TreasuryVoteIntervalMultiplier) {
+			s, _ := standalone.CalculateTSpendWindowStart(exp,
+				b.chainParams.TreasuryVoteInterval,
+				b.chainParams.TreasuryVoteIntervalMultiplier)
+
+			str := fmt.Sprintf("block contains TSpend "+
+				"transaction (%v) that is outside "+
+				"of the valid window: height %v "+
+				"start %v expiry %v",
+				stx.Hash(), block.Height(), s, exp)
+			return ruleError(ErrInvalidTSpendWindow, str)
+		}
+
+		tspends = append(tspends, stx)
+
+		// Verify this TSpend hash has not been included in a
+		// prior block.
+		err := b.checkTSpendExists(block, prevNode, stx)
```

Comment:
I'm confused: Isn't this trivially false (no other TSPends exist) considering a TSpend needs to encode the block height where they may be included? 

I guess it could happen with vanishing probability that a tspend tx hashes to the *exact same hash* as another previous tspend, but how would that be a problem (other than being an issue for the utxo set/utxoview to deal with, such as disallowing new tspends if the previous one is not fully spent).

Reply:
Ha! that is not what this is for. What we are preventing here is the SAME `TSPEND` to be mined a (or a few) TVIs later and prevent replays. A `TSPEND` is by definition valid for the `Expiry` duration and therefore we must reject additional inclusions.

Reply:
Ok, I _think_ I convinced myself this actually **is** needed and you can't rely just on utxo set semantics to ensure the same tspend can't be mined.

Basically, you could arrange all receivers to spend the TSPEND outputs immediately (after the maturity) and on the next TVI (but still inside the expiry) you could resend the same TSPEND tx.